File: ios/App/App/FolioleCompanionSyncGroupSigning.swift
import Capacitor
import Foundation

extension FolioleCompanionSyncPlugin {
    @objc func signCompanionSyncRequest(_ call: CAPPluginCall) {
        DispatchQueue.global(qos: .userInitiated).async {
            do { call.resolve(try self.signSyncGroupRequest(call)) }
            catch { call.reject("Failed to sign Sync Group request: \(error.localizedDescription)", nil, error) }
        }
    }

    private func signSyncGroupRequest(_ call: CAPPluginCall) throws -> [String: Any] {
        let contract = try FolioleCompanionContractStore().networkContract()
        func value(_ name: String) throws -> String {
            guard let key = contract.signatureRequestKeys[name], let value = call.getString(key), !value.isEmpty else {
                throw NSError(domain: "FolioleCompanionSyncGroupSigning", code: 1,
                              userInfo: [NSLocalizedDescriptionKey: "\(name)_required"])
            }
            return value
        }
        let groupId = try value("syncGroupId")
        let credential = try groupData.request("load_current_credential", ["group_id": groupId])
        guard let deviceId = credential["device_id"] as? String,
              let workgroupKey = credential["workgroup_key"] as? String else {
            throw NSError(domain: "FolioleCompanionSyncGroupSigning", code: 2,
                          userInfo: [NSLocalizedDescriptionKey: "sync_group_current_credential_missing"])
        }
        return try FolioleCompanionSignedClientRequests.prepare(
            body: call.getString(contract.signatureRequestKeys["body"]!),
            bodyHash: try value("bodyHash"), endpointUrl: try value("endpointUrl"), groupId: groupId,
            method: try value("method").uppercased(), nonce: try value("nonce"),
            path: try value("pathWithQuery"), timestamp: try value("timestamp"),
            deviceId: deviceId, workgroupKey: workgroupKey
        )
    }
}

File: android/app/src/main/java/com/foliole/android/FolioleCompanionSyncGroupSigning.java
package com.foliole.android;

import android.content.Context;

import com.getcapacitor.JSObject;
import com.getcapacitor.PluginCall;

import org.json.JSONObject;

final class FolioleCompanionSyncGroupSigning {
    private FolioleCompanionSyncGroupSigning() {}

    static JSObject sign(Context context, PluginCall call) throws Exception {
        String groupId = required(call, "sync_group_id");
        String endpointUrl = required(call, "endpoint_url");
        String method = required(call, "method");
        String path = required(call, "path_with_query");
        FolioleCompanionCurrentGroupCredential credential = FolioleCompanionCurrentGroupCredential.load(groupId);
        String body = call.getString("body");
        if (body != null) {
            JSONObject headers = new JSONObject().put("X-Sync-Group-Id", groupId)
                .put("X-Device-Id", credential.deviceId);
            FolioleCompanionWorkgroupHttp.PreparedRequest prepared = FolioleCompanionWorkgroupHttp.prepareWithKey(
                context, endpointUrl + path, method, headers, body, credential.workgroupKey);
            return new JSObject().put("body", prepared.body).put("headers", prepared.headers);
        }
        String timestamp = required(call, "timestamp");
        String nonce = required(call, "nonce");
        String bodyHash = required(call, "body_hash");
        String canonical = method.toUpperCase() + "\n" + path + "\n" + timestamp + "\n" + nonce + "\n" + bodyHash;
        return new JSObject().put("headers", new JSONObject()
            .put("X-Device-Id", credential.deviceId)
            .put("X-Nonce", nonce)
            .put("X-Signature", FolioleCompanionSyncGroupHmac.sign(credential.workgroupKey, canonical))
            .put("X-Timestamp", timestamp));
    }

    private static String required(PluginCall call, String key) {
        String value = call.getString(key);
        if (value == null || value.trim().isEmpty()) throw new IllegalArgumentException(key + "_required");
        return value.trim();
    }
}
